libs/core/tests/unit/mutex.cpp

```diff
@@ -116,4 +119,77 @@ TEST(DebugMutex, DISABLED_MultiThreadDeadlock2)
   threads.clear();
 }
 
+TEST(DebugMutex, CorrectRecursive)
+{
+  fetch::DeadlockHandler::ThrowOnDeadlock();
+  RMutex m;
+
+  EXPECT_TRUE(m.try_lock());
+  EXPECT_TRUE(m.try_lock());
+  EXPECT_TRUE(m.try_lock());
+  EXPECT_TRUE(m.try_lock());
+  m.unlock();
+  m.unlock();
+  m.unlock();
+  m.unlock();
+
+  std::string rv;
+
+  auto f = [&m, &rv](char c) {
+    m.lock();
+    m.lock();
+    m.lock();
+    m.lock();
+
+    rv += c;
+    rv += c;
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(250));
+
+    m.unlock();
+    m.unlock();
+
+    rv += c;
+    rv += c;
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(250));
+
+    m.unlock();
+    m.unlock();
+  };
+
+  std::vector<std::thread> threads;
+  threads.emplace_back(f, 'a');
+  threads.emplace_back(f, 'b');
+
+  threads[0].join();
+  threads[1].join();
+
+  ASSERT_TRUE(rv == "aaaabbbb" || rv == "bbbbaaaa");
+}
+
+TEST(DebugMutex, IncorrectRecursive)
+{
+  fetch::DeadlockHandler::ThrowOnDeadlock();
+  RMutex m;
+
+  std::vector<std::thread> threads;
+
+  threads.emplace_back([&m] {
+    FETCH_LOCK(m);
+    FETCH_LOCK(m);
+    std::this_thread::sleep_for(std::chrono::seconds(4));
```

Comment:
This test can fail based on race conditions and so is flaky. It is not obvious why there is a 4 second wait and will not scale to longer timeout durations. It also unneccessarily incresases our CI time to have sleeps when they could be avoided. I would suggest rewriting it to address these concerns

Reply:
4 seconds are chosen as a time greater than assumed deadlock threshold of 2s. You've mentioned above that this should be 40m (0_o) which definitely exceeds _any_ CI time.